maps/snail_mode.go

```diff
@@ -0,0 +1,196 @@
+package maps
+
+import (
+	"github.com/BattlesnakeOfficial/rules"
+)
+
+type SnailModeMap struct{}
+
+// init registers this map in the global registry.
+func init() {
+	globalRegistry.RegisterMap("snail_mode", SnailModeMap{})
+}
+
+// ID returns a unique identifier for this map.
+func (m SnailModeMap) ID() string {
+	return "snail_mode"
+}
+
+// Meta returns the non-functional metadata about this map.
+func (m SnailModeMap) Meta() Metadata {
+	return Metadata{
+		Name:        "Snail Mode",
+		Description: "Snakes leave behind a trail of hazards",
+		Author:      "coreyja and jlafayette",
+		Version:     1,
+		MinPlayers:  1,
+		MaxPlayers:  16,
+		BoardSizes:  OddSizes(rules.BoardSizeSmall, rules.BoardSizeXXLarge),
+	}
+}
+
+// SetupBoard here is pretty 'standard' and doesn't do any special setup for this game mode
+func (m SnailModeMap) SetupBoard(initialBoardState *rules.BoardState, settings rules.Settings, editor Editor) error {
+	rand := settings.GetRand(0)
+
+	if len(initialBoardState.Snakes) > int(m.Meta().MaxPlayers) {
+		return rules.ErrorTooManySnakes
+	}
+
+	snakeIDs := make([]string, 0, len(initialBoardState.Snakes))
+	for _, snake := range initialBoardState.Snakes {
+		snakeIDs = append(snakeIDs, snake.ID)
+	}
+
+	tempBoardState := rules.NewBoardState(initialBoardState.Width, initialBoardState.Height)
+	err := rules.PlaceSnakesAutomatically(rand, tempBoardState, snakeIDs)
+	if err != nil {
+		return err
+	}
+
+	// Copy snakes from temp board state
+	for _, snake := range tempBoardState.Snakes {
+		editor.PlaceSnake(snake.ID, snake.Body, snake.Health)
+	}
+
+	return nil
+}
+
+// hash takes in a Point and the height of the board, and returns a unique integer to be used in maps
+// Convert the given point into an integer that can be used as a map key
+// This is takes the X and Y coordinates and combines them into a single integer
+// that uniquely identifies the point given the current board size
+func hash(point rules.Point, height int) int {
+	return point.X + point.Y*height
+}
```

Comment:
It's probably not worth changing, but you can actually use a `rules.Point` as a map key! I think a few other maps use this to look up tables of locations for a given WxH board.

Reply:
I didn't know that was possible to use a struct as a map key - thanks!